src/osd/PG.cc

```diff
@@ -7251,22 +7237,26 @@ boost::statechart::result PG::RecoveryState::Active::react(const AdvMap& advmap)
   }
 
   bool need_publish = false;
+  pg->_update_calc_stats(); // Too bad this is called again from publish_stats_to_osd()
+  int64_t degraded = pg->info.stats.stats.sum.num_objects_degraded;
+  if (!pg->state_test(PG_STATE_DEGRADED) && degraded) {
+    need_publish = true;
+    pg->state_set(PG_STATE_DEGRADED);
+  } else if (pg->state_test(PG_STATE_DEGRADED) && !degraded) {
+    need_publish = true;
+    pg->state_clear(PG_STATE_DEGRADED);
+  }
+
   /* Check for changes in pool size (if the acting set changed as a result,
    * this does not matter) */
   if (advmap.lastmap->get_pg_size(pg->info.pgid.pgid) !=
       pg->get_osdmap()->get_pg_size(pg->info.pgid.pgid)) {
     if (pg->get_osdmap()->get_pg_size(pg->info.pgid.pgid) <= pg->actingset.size()) {
       pg->state_clear(PG_STATE_UNDERSIZED);
-      if (pg->needs_recovery()) {
-	pg->state_set(PG_STATE_DEGRADED);
-      } else {
-	pg->state_clear(PG_STATE_DEGRADED);
-      }
     } else {
       pg->state_set(PG_STATE_UNDERSIZED);
-      pg->state_set(PG_STATE_DEGRADED);
     }
-    need_publish = true; // degraded may have changed
+    need_publish = true; // undersized may have changed
```

Comment:
...then this would just need to set need_publish, and the update_calc_stats would update the DEGRADED flag?